[idl_parser] Mark typefield as deprecated

liuliu opened the pull request:
For union types, the deprecated field is not propagated to its typefield, therefore, the generated code will be erroneous because the typefield will be emitted.

This PR simply propagated deprecated field to the typefield too.

liuliu commented:
@googlebot I signed it!

dbaileychess commented:
The later option sounds better. Thanks!

liuliu commented:
Hi, @dbaileychess , added test in https://github.com/google/flatbuffers/pull/5958/commits/570f4ce6fbc31b3c6a31c2cb7913f997fa05673d

Tried on google/master, it will fail this test (because explicit check for c_type field from parser will not be marked as deprecated). With this PR, the value will be toggled properly.

dbaileychess commented:
Generally looks good. I like that the tests are separate, but could we just add new tables/unions to the evolution_test/evolution_v1.fbs? That way it is a little more condensed.

I think it should be doable without effect the evolution tests.

liuliu commented:
Do you mean I mark a union field in evolution_v2.fbs as deprecated, and have a separate test to check typefield->deprecated for that schema?

dbaileychess commented:
I would just add new union field to the root table in evolution_v1 and then deprecated it in v2. And then convert your tests to work off of those schemas.

This way most of the boilerplate of the check in (generated files, new includes, etc.) is not needed.

liuliu commented:
Updated. Note that evolution_v1_generated.h and evolution_v2_generated.h updated accordingly. Because it hasn't updated for a while, the naming changed from `Union::TableB` to `Union::Union_TableB`. Updated evolution test accordingly.

Let me know what do you think of this update. Thanks!

liuliu commented on tests/evolution_test/evolution_v1_generated.h:
```diff
@@ -42,24 +45,24 @@ inline const char * const *EnumNamesEnum() {
-    Union::NONE,
-    Union::TableA,
-    Union::TableB
+    Union_NONE,
```
I honestly don't know what happened, I simply run:
```
bazel run :flatc -- --cpp -o ~/workspace/flatbuffers/tests/evolution_test ~/workspace/flatbuffers/tests/evolution_test/evolution_v1.fbs
bazel run :flatc -- --cpp -o ~/workspace/flatbuffers/tests/evolution_test ~/workspace/flatbuffers/tests/evolution_test/evolution_v2.fbs
```
and these updated generated. Should I use a different set of options?

dbaileychess replied:
The default `test/generate_code.sh` script should suffice. I just checked it and the line to generated it was removed here for some reason: https://github.com/google/flatbuffers/commit/44bf719883248a392e04d9cebecdedb00cdd393c#diff-095ca9891b22d443d2181ad26d248555L22

You can use this line for now:

`flatc --cpp --scoped-enums -o evolution_test ./evolution_test/evolution_v*.fbs`

I'll look into re adding this line again.

liuliu commented:
Thanks! Updated with `--scoped-enums`.

dbaileychess commented:
Looks good. I will merge when the CI is complete!
